Source/OCMock/OCMFunctions.m

```diff
@@ -397,35 +397,58 @@ SEL OCMOriginalSelectorForAlias(SEL selector)
 
 void OCMSetAssociatedMockForClass(OCClassMockObject *mock, Class aClass)
 {
-    if((mock != nil) && (objc_getAssociatedObject(aClass, OCMClassMethodMockObjectKey) != nil))
-        [NSException raise:NSInternalInconsistencyException format:@"Another mock is already associated with class %@", NSStringFromClass(aClass)];
-    objc_setAssociatedObject(aClass, OCMClassMethodMockObjectKey, mock, OBJC_ASSOCIATION_ASSIGN);
+    // Prevent the associated mock from being destroyed between the time when
+    // OCMRetainAssociatedMockForClass gets the mock, and retains it on a different thread.
+    @synchronized(OCMClassMethodMockObjectKey)
+    {
+        if((mock != nil) && (objc_getAssociatedObject(aClass, OCMClassMethodMockObjectKey) != nil))
+            [NSException raise:NSInternalInconsistencyException format:@"Another mock is already associated with class %@", NSStringFromClass(aClass)];
+        objc_setAssociatedObject(aClass, OCMClassMethodMockObjectKey, mock, OBJC_ASSOCIATION_ASSIGN);
+    }
 }
 
-OCClassMockObject *OCMGetAssociatedMockForClass(Class aClass, BOOL includeSuperclasses)
+OCClassMockObject *OCMRetainAssociatedMockForClass(Class aClass, BOOL includeSuperclasses)
 {
-    OCClassMockObject *mock = nil;
-    do
+    // See OCMSetAssociatedMockForClass regarding @synchronized.
+    @synchronized(OCMClassMethodMockObjectKey)
     {
-        mock = objc_getAssociatedObject(aClass, OCMClassMethodMockObjectKey);
-        aClass = class_getSuperclass(aClass);
+        do
+        {
+
+            OCClassMockObject *mock = objc_getAssociatedObject(aClass, OCMClassMethodMockObjectKey);
+            if(mock != nil)
+            {
+                return [mock retain];
+            }
+            aClass = class_getSuperclass(aClass);
+        }
+        while((aClass != nil) && includeSuperclasses);
     }
-    while((mock == nil) && (aClass != nil) && includeSuperclasses);
-    return mock;
+    return nil;
 }
 
 static NSString *const OCMPartialMockObjectKey = @"OCMPartialMockObjectKey";
 
 void OCMSetAssociatedMockForObject(OCClassMockObject *mock, id anObject)
 {
-    if((mock != nil) && (objc_getAssociatedObject(anObject, OCMPartialMockObjectKey) != nil))
-        [NSException raise:NSInternalInconsistencyException format:@"Another mock is already associated with object %@", anObject];
-    objc_setAssociatedObject(anObject, OCMPartialMockObjectKey, mock, OBJC_ASSOCIATION_ASSIGN);
+    // Prevent the associated mock from being destroyed between the time when
+    // OCMRetainAssociatedMockForClass gets the mock, and retains it on a different thread.
+    @synchronized(OCMPartialMockObjectKey)
```

Comment:
I don't follow this reasoning. We're using `OBJC_ASSOCIATION_ASSIGN` which does not retain the associated object. So, even if the object is set while another thread is in between lines 449 and 450, this shouldn't make a difference, right?

Reply:
yeah.. I had managed to convince myself that there was some level of safety here, but I think you are correct. Given that it's an assign the pointer could be dangling at any time, and the @synchronized blocks aren't going to stop it from being released.

Reply:
I guess this means I can close this PR, right?

Reply:
Yeah.. we can close this one. Thanks.